src/main/java/org/veupathdb/service/eda/ss/model/reducer/FormattedTabularRecordStreamer.java

```diff
@@ -0,0 +1,124 @@
+package org.veupathdb.service.eda.ss.model.reducer;
+
+import org.veupathdb.service.eda.ss.model.Entity;
+import org.veupathdb.service.eda.ss.model.variable.VariableValueIdPair;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Provides a stream of variable value records extracted from a stream of ID indexes and a collection of
+ * ID index, variable value pairs. The resulting stream provides formatted tuples of strings containing entity IDs,
+ * ancestor IDs and all associated variable values.
+ */
+public class FormattedTabularRecordStreamer implements Iterator<List<String>> {
+  private List<ValueStream> valuePairStreams;
+  private ValueStream<List<Long>> ancestorStream;
+  private Iterator<Long> idIndexStream;
+  private Long currentIdIndex;
+  private Entity entity;
+
+  /**
+   * Constructs an instance, which provides a stream of string-formatted records. The stream is composed of all IDs
+   * in the idIndex stream, corresponding ancestor IDs corresponding values in the
+   * {@link FormattedTabularRecordStreamer#valuePairStreams}
+   *
+   * @param valuePairStreams Streams of pairs containing entity ID indexes and corresponding values for their respective
+   *                         variables.
+   * @param idIndexStream    Stream of ID indexes, indicating which entity records to output.
+   * @param ancestorStream   Stream of ancestors, used to merge in ancestors of target output entities in output records.
+   * @param entity           Entity type of records which will be output.
+   */
+  public FormattedTabularRecordStreamer(List<Iterator<VariableValueIdPair<?>>> valuePairStreams,
+                                        Iterator<Long> idIndexStream,
+                                        Iterator<VariableValueIdPair<List<Long>>> ancestorStream,
+                                        Entity entity) {
+    this.valuePairStreams = valuePairStreams.stream()
```

Comment:
its a little confusing that the instance var and param have same name but are diff types.